### ISSUE TITLE
ONNX: pytorch convert to ONNX model, the result of inference is wrong

### ISSUE THREAD

**Issue by ChisenZhang** [pytorch/pytorch#21455]
## ❓ Questions and Help
I use the api of torch.onnx save Faceboxes model to onnx model. checked the model of onnx by onnx.check, the total stage of torch model to onnx model was not error or warning message. But the inference result of ONNX model is wrong. the version of pytorch is 1.1.0, the onnxruntime is 0.4, the onnx is 1.5. A classification ONNX model of converted from TF model is normal.  what's the reason maybe?  What's the version of ONNX is supported by pytorch? Thanks.

cc @BowenBao @neginraoof

**Comment by fmassa** [comment]
Please provide an example of the model architecture, as there is not enough information here so that we can help you further debug this issue.

**Comment by ChisenZhang** [comment]
> Please provide an example of the model architecture, as there is not enough information here so that we can help you further debug this issue.

this is the model architecture of faceboxes:
`import torch
import torch.nn as nn
import torch.nn.functional as F


class BasicConv2d(nn.Module):

    def __init__(self, in_channels, out_channels, **kwargs):
        super(BasicConv2d, self).__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, bias=False, **kwargs)
        self.bn = nn.BatchNorm2d(out_channels, eps=1e-5)

    def forward(self, x):
        x = self.conv(x)
        x = self.bn(x)
        return F.relu(x, inplace=True)


class Inception(nn.Module):

  def __init__(self):
    super(Inception, self).__init__()
    self.branch1x1 = BasicConv2d(128, 32, kernel_size=1, padding=0)
    self.branch1x1_2 = BasicConv2d(128, 32, kernel_size=1, padding=0)
    self.branch3x3_reduce = BasicConv2d(128, 24, kernel_size=1, padding=0)
    self.branch3x3 = BasicConv2d(24, 32, kernel_size=3, padding=1)
    self.branch3x3_reduce_2 = BasicConv2d(128, 24, kernel_size=1, padding=0)
    self.branch3x3_2 = BasicConv2d(24, 32, kernel_size=3, padding=1)
    self.branch3x3_3 = BasicConv2d(32, 32, kernel_size=3, padding=1)
  
  def forward(self, x):
    branch1x1 = self.branch1x1(x)
    
    branch1x1_pool = F.avg_pool2d(x, kernel_size=3, stride=1, padding=1)
    branch1x1_2 = self.branch1x1_2(branch1x1_pool)
    
    branch3x3_reduce = self.branch3x3_reduce(x)
    branch3x3 = self.branch3x3(branch3x3_reduce)
    
    branch3x3_reduce_2 = self.branch3x3_reduce_2(x)
    branch3x3_2 = self.branch3x3_2(branch3x3_reduce_2)
    branch3x3_3 = self.branch3x3_3(branch3x3_2)
    
    outputs = [branch1x1, branch1x1_2, branch3x3, branch3x3_3]
    return torch.cat(outputs, 1)


class CRelu(nn.Module):

  def __init__(self, in_channels, out_channels, **kwargs):
    super(CRelu, self).__init__()
    self.conv = nn.Conv2d(in_channels, out_channels, bias=False, **kwargs)
    self.bn = nn.BatchNorm2d(out_channels, eps=1e-5)
  
  def forward(self, x):
    x = self.conv(x)
    x = self.bn(x)
    x = torch.cat([x, -x], 1)
    x = F.relu(x, inplace=True)
    return x
    
    
class FaceBoxes(nn.Module):

  def __init__(self, phase, size, num_classes):
    super(FaceBoxes, self).__init__()
    self.phase = phase
    self.num_classes = num_classes
    self.size = size
    
    self.conv1 = CRelu(3, 24, kernel_size=7, stride=4, padding=3)
    self.conv2 = CRelu(48, 64, kernel_size=5, stride=2, padding=2)
    
    self.inception1 = Inception()
    self.inception2 = Inception()
    self.inception3 = Inception()
    
    self.conv3_1 = BasicConv2d(128, 128, kernel_size=1, stride=1, padding=0)
    self.conv3_2 = BasicConv2d(128, 256, kernel_size=3, stride=2, padding=1)
    
    self.conv4_1 = BasicConv2d(256, 128, kernel_size=1, stride=1, padding=0)
    self.conv4_2 = BasicConv2d(128, 256, kernel_size=3, stride=2, padding=1)
    
    self.loc, self.conf = self.multibox(self.num_classes)
    
    if self.phase == 'test':
        self.softmax = nn.Softmax(dim=-1)

    if self.phase == 'train':
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                if m.bias is not None:
                    nn.init.xavier_normal_(m.weight.data)
                    m.bias.data.fill_(0.02)
                else:
                    m.weight.data.normal_(0, 0.01)
            elif isinstance(m, nn.BatchNorm2d):
                m.weight.data.fill_(1)
                m.bias.data.zero_()

  def multibox(self, num_classes):
    loc_layers = []
    conf_layers = []
    loc_layers += [nn.Conv2d(128, 21 * 4, kernel_size=3, padding=1)]
    conf_layers += [nn.Conv2d(128, 21 * num_classes, kernel_size=3, padding=1)]
    loc_layers += [nn.Conv2d(256, 1 * 4, kernel_size=3, padding=1)]
    conf_layers += [nn.Conv2d(256, 1 * num_classes, kernel_size=3, padding=1)]
    loc_layers += [nn.Conv2d(256, 1 * 4, kernel_size=3, padding=1)]
    conf_layers += [nn.Conv2d(256, 1 * num_classes, kernel_size=3, padding=1)]
    return nn.Sequential(*loc_layers), nn.Sequential(*conf_layers)
    
  def forward(self, x):
  
    sources = list()
    loc = list()
    conf = list()
    detection_dimension = list()

    x = self.conv1(x)
    x = F.max_pool2d(x, kernel_size=3, stride=2, padding=1)
    x = self.conv2(x)
    x = F.max_pool2d(x, kernel_size=3, stride=2, padding=1)
    x = self.inception1(x)
    x = self.inception2(x)
    x = self.inception3(x)
    detection_dimension.append(x.shape[2:])
    sources.append(x)
    x = self.conv3_1(x)
    x = self.conv3_2(x)
    detection_dimension.append(x.shape[2:])
    sources.append(x)
    x = self.conv4_1(x)
    x = self.conv4_2(x)
    detection_dimension.append(x.shape[2:])
    sources.append(x)
    
    detection_dimension = torch.tensor(detection_dimension, device=x.device)

    for (x, l, c) in zip(sources, self.loc, self.conf):
        loc.append(l(x).permute(0, 2, 3, 1).contiguous())
        conf.append(c(x).permute(0, 2, 3, 1).contiguous())
        
    loc = torch.cat([o.view(o.size(0), -1) for o in loc], 1)
    conf = torch.cat([o.view(o.size(0), -1) for o in conf], 1)

    if self.phase == "test":
      output = (loc.view(loc.size(0), -1, 4),
                self.softmax(conf.view(-1, self.num_classes)),
                detection_dimension)
    else:
      output = (loc.view(loc.size(0), -1, 4),
                conf.view(conf.size(0), -1, self.num_classes),
                detection_dimension)
  
    return output
`

**Comment by ChisenZhang** [comment]
> Please provide an example of the model architecture, as there is not enough information here so that we can help you further debug this issue.

if you have some time, can you help me to analysis this problem?

**Comment by Wheest** [comment]
I've encountered a relatively similar issue, and could perhaps bring some clarity (though no solution).

Possibly a mistake on my part, but here goes:

I think the model in question is InceptionV3: https://github.com/pytorch/vision/blob/master/torchvision/models/inception.py

I've been exporting the PyTorch Onnx model to a third party framework.  Between frameworks, the output of every layer until the first avgpool is correct.

When I export the model to Onnx, the [first AvgPool layer](https://github.com/pytorch/vision/blob/master/torchvision/models/inception.py#L229) decomposes to a [Pad](https://github.com/onnx/onnx/blob/master/docs/Operators.md#pad), followed by an [AveragePool](https://github.com/onnx/onnx/blob/master/docs/Operators.md#AveragePool) (trivial to collapse, but not done by default in my output).

As you can see in the InceptionV3 source code, this avgerage pool layer has pad set to 1.  However, examining the 'pads' attribute of the Pad op, we can see that it is `(0, 1, 0, 1)`.

This means that (in Onnx) the inputs to the next [Concat](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Concat) layer are of shape:

```
600: [1, 64, 25, 25]
606: [1, 64, 25, 25]
615: [1, 96, 25, 25]
620: [1, 32, 24, 24] # output of average pool
```

However, if I manually set the padding to `(1, 1, 1, 1)` to make the Concat operation work, the output of that layer is incorrect.

Perhaps there's an edge case in the PyTorch Onnx exporter?  

For the record, I am determining the shape of the output of average pool with:

```
if ceil_mode == 1:
    out_h = math.ceil(((in_h + pad_hl + pad_hr - kernel_size) / stride) + 1)
    out_w = math.ceil(((in_w + pad_wl + pad_wr - kernel_size) / stride) + 1)
else:
    out_h = int((in_h + pad_hl + pad_hr - kernel_size) / stride + 1)
    out_w = int((in_w + pad_wl + pad_wr - kernel_size) / stride + 1)
```

Where in this case `in_h, in_w = 25, 25`, `kernel_size = 3` and `stride = 1`

**Comment by jagin** [comment]
Hi @ChisenZhang,

I stumbled on the same problem. I retrained the FaceBoxes.PyTorch and was trying to convert the net to the ONNX format without any problem with the following code:

```
import torch

from models.faceboxes import FaceBoxes

img_dim = (720, 1280)
rgb_mean = (104, 117, 123) # bgr order
num_classes = 2
model_path = "weights/Final_FaceBoxes.pth"
onnx_path = "weights/Final_FaceBoxes.onnx"

# Create the model and load the weights
model = FaceBoxes('test', img_dim, num_classes)

state_dict = torch.load(model_path)
model.load_state_dict(state_dict, strict=False)

# Create dummy input
dummy_input = torch.rand(1, 3, img_dim[0], img_dim[1])

# Define input / output names
input_names = ["actual_input_1"]
output_names = ["output1", "353"]

# Convert the PyTorch model to ONNX
torch.onnx.export(model,
                  dummy_input,
                  onnx_path,
                  verbose=True,
                  input_names=input_names,
                  output_names=output_names)
```

**but the inference was the mess.**

The hint on how to resolve this problem is in [stackoverflow](https://stackoverflow.com/questions/44230907/keyerror-unexpected-key-module-encoder-embedding-weight-in-state-dict) and in [train.py](https://github.com/zisianw/FaceBoxes.PyTorch/blob/9bc5811fe8c409a50c9f23c6a770674d609a2c3a/train.py#L54) (line 54-66).

I changed my ONNX converter to remove *module* prefix in the following way:

```
import torch

from models.faceboxes import FaceBoxes

img_dim = (720, 1280)
rgb_mean = (104, 117, 123)  # bgr order
num_classes = 2
model_path = "weights/Final_FaceBoxes.pth"
onnx_path = "weights/Final_FaceBoxes.onnx"

# Create the model and load the weights
model = FaceBoxes('test', img_dim, num_classes)

state_dict = torch.load(model_path)
# create new OrderedDict that does not contain `module.`
from collections import OrderedDict
new_state_dict = OrderedDict()
for k, v in state_dict.items():
    head = k[:7]
    if head == 'module.':
        name = k[7:]  # remove `module.`
    else:
        name = k
    new_state_dict[name] = v
model.load_state_dict(new_state_dict)

# Create dummy input
dummy_input = torch.rand(1, 3, img_dim[0], img_dim[1])

# Define input / output names
input_names = ["actual_input_1"]
output_names = ["output1", "353"]

# Convert the PyTorch model to ONNX
torch.onnx.export(model,
                  dummy_input,
                  onnx_path,
                  verbose=True,
                  input_names=input_names,
                  output_names=output_names)
```

and the inference was perfect (I was using the onnx model with OpenCV).

I hope it helps. Happy coding :)

**Comment by lecheel** [comment]
can share the openCV code for onnx model how perfect is it

**Comment by garymm** [comment]
@ChisenZhang is this still an issue with the nightly version of PyTorch and onnxruntime 1.9.0?

If so, can you please provide a minimal repro, ideally in a colab notebook?

**Comment by CaichaoGitHub** [comment]
same problem , i convert the  [model](https://github.com/david-wb/gaze-estimation) to onnx ,and i predicted using onnxruntime, the result is different with the torch model

**Comment by thiagocrepaldi** [comment]
@CaichaoGitHub we need the full script, not just a URL for the model you mentioned. If you are unable to provide it, we will close this issue

**Comment by shubhambhokare1** [comment]
Hi @CaichaoGitHub

We’ve gone ahead and closed this issue because it is stale.
If you still believe this issue is relevant, please feel free to reopen the issue and we will triage it as necessary. Please specify in a comment any updated information you may have so that we can address it effectively. We encourage you to try the latest pytorch-preview (nightly) version to see if it has resolved the issue.

Thanks,
ONNX Converter team
